### PR TITLE
(WIP) YONK-529 Discussions web fragments prototype

### PR DESCRIPTION
**Pull request by bogdanl**
Branches from `open-craft:carlio/OC-2714-web-fragments2`, which is not merged to master yet, hence the big diff.

This is a WIP attempt for allowing the discussion tab in Apros to be rendered via web fragments, so that the course discussion XBlock is not involved anymore. It makes use of changes to `jquery-xblock-2.js` in the Apros repository, in order to display web fragments similarly to XBlocks.

The difficult part is proving to be making RequireJS work both on the standalone view and via `jquery-xblock`, specifically the require/define imports while adding JS files and inline JS blocks dynamically.

**Testing instructions**:
1. Apply [this patch](https://github.com/edx-solutions/edx-platform/files/1308342/bogdan-jquery-xblock-fragments.diff.zip) to Apros's `jquery-xblock` that allows web fragments to be displayed via `jquery-xblock`. Please note that that's also WIP and it's likely it will need more work.
2. Add a section named `DISCUSSION_TAB` to a course and in Apros go to the Discussions tab. 
3. Check the JS console for the import error messages.

**Reviewers**:
- [ ] @e-kolpakov

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @bogdanl! I've created [OSPR-1892](https://openedx.atlassian.net/browse/OSPR-1892) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx-solutions/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

File: lms/djangoapps/discussion/urls.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
     url(r'^(?P<discussion_id>[\w\-.]+)/threads/(?P<thread_id>\w+)$', 'single_thread', name='single_thread'),
     url(r'^(?P<discussion_id>[\w\-.]+)/inline$', 'inline_discussion', name='inline_discussion'),
     url(
-        r'discussion_board_fragment_view$',
+        r'discussion_board_fragment_view',
```
Comment (e-kolpakov) [inline lms/djangoapps/discussion/urls.py]: @bogdanl As far as I know, Django's url definitions does not affect presense or absense of the querystring. Could you please explain why this change is necessary?

File: lms/djangoapps/discussion/views.py [hunk 2]
```diff
@@ -671,9 +691,20 @@ def render_to_fragment(self, request, course_id=None, discussion_id=None, thread
...
+
+            if request.GET.get('format', 'html') == 'json':
+                fragment.add_javascript_url(staticfiles_storage.url('common/js/vendor/require.js'))
+                fragment.add_javascript_url(staticfiles_storage.url('lms/js/require-config.js.js'))
+
             if not settings.REQUIRE_DEBUG:
-                fragment.add_javascript_url(staticfiles_storage.url('discussion/js/discussion_board_factory.js'))
+                fragment.add_javascript_url(
+                    staticfiles_storage.url(
+                        staticfiles_storage.url('discussion/js/discussion_board_factory.js')
```
Comment (e-kolpakov) [inline lms/djangoapps/discussion/views.py]: @bogdanl Why double `staticfiles_storage.url`?

File: lms/djangoapps/discussion/views.py [hunk 3]
```diff
@@ -671,9 +691,20 @@ def render_to_fragment(self, request, course_id=None, discussion_id=None, thread
 
             fragment = Fragment(html)
             self.add_fragment_resource_urls(fragment)
-            fragment.add_javascript(inline_js)
+
+            if request.GET.get('format', 'html') == 'json':
+                fragment.add_javascript_url(staticfiles_storage.url('common/js/vendor/require.js'))
+                fragment.add_javascript_url(staticfiles_storage.url('lms/js/require-config.js.js'))
```
Comment (e-kolpakov) [inline lms/djangoapps/discussion/views.py]: @bogdanl This might be the place that causes you trouble. `require-config.js` does not have discussion JS definitions, if I recall right. You might want to introduce a separate, discussion-specific config and add it here as well.

A better approach would be to make discussions always use require.js, but that's probably outside of the scope of current task.

File: lms/djangoapps/discussion/urls.py (hunk 1 above)
Comment (bogdanl) [reply lms/djangoapps/discussion/urls.py]: Without the $ at the end, any call to the `discussion_board_fragment_view` would match the root URL handler in that urlconf with querystring appended. Actually the underlying problem might be the missing $ in the root handler, or a combination of the two. Like this it will match both URLs correctly though. I will check again to make sure that's the proper fix.

File: lms/djangoapps/discussion/views.py (hunk 2 above)
Comment (bogdanl) [reply lms/djangoapps/discussion/views.py]: Oh sorry about that, rushed commit – thanks for pointing it out.

File: lms/djangoapps/discussion/urls.py (hunk 1 above)
Comment (bogdanl) [reply lms/djangoapps/discussion/urls.py]: If we don't remove the trailing `$`, the resolver won't match `.../discussion_board_fragment_view/` (with a trailing slash); nor it will match `.../discussion_board_fragment_view?format=json`. If we remove it, it will match cover the trailing slash scenario as well and all combinations of trailing slash with or without querystring will work.

Furthermore the `forum_form_discussion` handler needs a more restrictive regex, as at the moment it matches anything that's not matched in the previous handlers, causing unexpected side effects.

Comment (e-kolpakov) [reply lms/djangoapps/discussion/urls.py]: @bogdanl If that's the practical effect than you're probably right, but theoretically it shoudln't be the case:

* `discussion_board_fragment_view?format=json` - querystring is not matched by the url pattern
* `/discussion_board_fragment_view/` - if I'm not mistaking trailing slashes must be automatically appended via 302 client redirect (there's a Django setting to do so, and I believe it is enabled).

So, probably we were looking at the problem from the wrong angle - if you just add a trailing slash *and* `$` it might just work.

File: lms/djangoapps/discussion/views.py (hunk 3 above)
Comment (bogdanl) [reply lms/djangoapps/discussion/views.py]: Aren't the discussion JS definitions being added with these two lines? 
`inline_js = render_to_string('discussion/discussion_board_js.template', context)`
`fragment.add_javascript(inline_js)`

Comment (e-kolpakov) [reply lms/djangoapps/discussion/views.py]: @bogdanl Yes, you're right (that's the effect of delaying the task for couple of months - everyone forgets everything).

Well, since discussions requireJS config is already included, but Apros fails to use it the only remaining options are:

1. Solve the problem using nginx - and it's not as bad asit sounds, since there are already [discussion-specific rules in it](https://github.com/mckinseyacademy/mcka_apros/blob/master/docs/mcka_apros_production#L45)
2. Somehow provide an overriding require.js config in Apros or jquery-xblock

**Comment by bogdanl** [comment]
I fixed a couple of path issues. Also please note
[this patch](https://github.com/edx-solutions/edx-platform/files/1324651/bogdan-mcka_apros-web-fragments.diff.zip) needs to be applied in mcka_apros.

**Comment by nedbat** [comment]
@bogdanl Is this ready to be reviewed? It still has (WIP) in the title.

**Comment by bradenmacdonald** [comment]
@nedbat  This was not supposed to be an OSPR, sorry. Please delete the OSPR ticket. This is WIP, yes, but done for the Business team, and we have handed it over to them. The JIRA ticket is YONK-529 and updates are on there.

CC @douglashall

**Comment by douglashall** [comment]
CC @smagoun
